File: android/core/data/src/main/java/app/pigeonsms/data/CapsuleCrypto.kt
package app.pigeonsms.data

import android.util.Base64
import java.security.MessageDigest
import java.security.SecureRandom
import javax.crypto.Cipher
import javax.crypto.SecretKeyFactory
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.PBEKeySpec
import javax.crypto.spec.SecretKeySpec

data class EncryptedCapsule(
    val ciphertext: String,
    val iv: String,
    val salt: String,
    val kdf: String = "pbkdf2-sha256-250000",
)

object CapsuleCrypto {
    private const val iterations = 250_000
    private val random = SecureRandom()

    fun encrypt(value: String, password: String): EncryptedCapsule {
        require(password.length >= 8) { "use at least 8 characters" }
        val salt = ByteArray(16).also(random::nextBytes)
        val iv = ByteArray(12).also(random::nextBytes)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.ENCRYPT_MODE, key(password, salt), GCMParameterSpec(128, iv))
        return EncryptedCapsule(
            ciphertext = encode(cipher.doFinal(value.toByteArray(Charsets.UTF_8))),
            iv = encode(iv),
            salt = encode(salt),
        )
    }

    fun decrypt(capsule: EncryptedCapsule, password: String): String {
        require(capsule.kdf == "pbkdf2-sha256-250000") { "unsupported checkpoint encryption" }
        val iv = decode(capsule.iv)
        val salt = decode(capsule.salt)
        val cipher = Cipher.getInstance("AES/GCM/NoPadding")
        cipher.init(Cipher.DECRYPT_MODE, key(password, salt), GCMParameterSpec(128, iv))
        return cipher.doFinal(decode(capsule.ciphertext)).toString(Charsets.UTF_8)
    }

    fun digest(ciphertext: String): String {
        val canonical = encode(decode(ciphertext))
        return MessageDigest.getInstance("SHA-256")
            .digest(canonical.toByteArray(Charsets.UTF_8))
            .joinToString("") { byte -> "%02x".format(byte.toInt() and 0xff) }
    }

    private fun key(password: String, salt: ByteArray): SecretKeySpec {
        val spec = PBEKeySpec(password.toCharArray(), salt, iterations, 256)
        val bytes = SecretKeyFactory.getInstance("PBKDF2WithHmacSHA256").generateSecret(spec).encoded
        spec.clearPassword()
        return SecretKeySpec(bytes, "AES")
    }

    private fun encode(bytes: ByteArray) = Base64.encodeToString(bytes, Base64.NO_WRAP)
    private fun decode(value: String) = Base64.decode(value, Base64.DEFAULT)
}

File: web/src/lib/capsuleCrypto.ts
const encoder = new TextEncoder()
const decoder = new TextDecoder()
const iterations = 250_000

function toBase64(bytes: Uint8Array): string {
  let binary = ''
  for (let index = 0; index < bytes.length; index += 0x8000) {
    binary += String.fromCharCode(...bytes.subarray(index, index + 0x8000))
  }
  return btoa(binary)
}

function fromBase64(value: string): Uint8Array<ArrayBuffer> {
  const binary = atob(value.replaceAll('-', '+').replaceAll('_', '/'))
  const bytes = new Uint8Array(binary.length)
  for (let index = 0; index < binary.length; index++) bytes[index] = binary.charCodeAt(index)
  return bytes
}

async function capsuleKey(password: string, salt: Uint8Array<ArrayBuffer>): Promise<CryptoKey> {
  const material = await crypto.subtle.importKey('raw', encoder.encode(password), 'PBKDF2', false, ['deriveKey'])
  return crypto.subtle.deriveKey(
    { name: 'PBKDF2', salt, iterations, hash: 'SHA-256' },
    material,
    { name: 'AES-GCM', length: 256 },
    false,
    ['encrypt', 'decrypt'],
  )
}

export interface EncryptedCapsule {
  ciphertext: string
  iv: string
  salt: string
  kdf: string
}

export async function encryptCapsule(value: unknown, password: string): Promise<EncryptedCapsule> {
  if (password.length < 8) throw new Error('use at least 8 characters')
  const salt = crypto.getRandomValues(new Uint8Array(16))
  const iv = crypto.getRandomValues(new Uint8Array(12))
  const data = await crypto.subtle.encrypt(
    { name: 'AES-GCM', iv },
    await capsuleKey(password, salt),
    encoder.encode(JSON.stringify(value)),
  )
  return {
    ciphertext: toBase64(new Uint8Array(data)),
    iv: toBase64(iv),
    salt: toBase64(salt),
    kdf: `pbkdf2-sha256-${iterations}`,
  }
}

export async function decryptCapsule<T>(capsule: EncryptedCapsule, password: string): Promise<T> {
  if (capsule.kdf !== `pbkdf2-sha256-${iterations}`) throw new Error('unsupported checkpoint encryption')
  const salt = fromBase64(capsule.salt)
  const iv = fromBase64(capsule.iv)
  const plain = await crypto.subtle.decrypt(
    { name: 'AES-GCM', iv },
    await capsuleKey(password, salt),
    fromBase64(capsule.ciphertext),
  )
  return JSON.parse(decoder.decode(plain)) as T
}

export async function capsuleDigest(ciphertext: string): Promise<string> {
  const canonical = toBase64(fromBase64(ciphertext))
  const digest = await crypto.subtle.digest('SHA-256', encoder.encode(canonical))
  return [...new Uint8Array(digest)].map((byte) => byte.toString(16).padStart(2, '0')).join('')
}
